ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala

```diff
@@ -0,0 +1,133 @@
+      case Some(value) =>
+        logger.debug(s"Cache hit for $key -> ${value.toString.take(100)}")
+        Some(value)
+      case None =>
+        logger.debug(s"Cache miss for $key ")
+        None
+    }
+
+  /** Update the cache synchronously.
+    *
+    * In face of multiple in-flight updates competing for the `key` (see [[putAsync()]]),
+    * this method updates the cache only if the to-be-inserted tuple is the most recent
+    * (i.e. it has `validAt` highest amongst the competing updates).
+    *
+    * @param key the key at which to update the cache
+    * @param validAt ordering discriminator for pending updates for the same key
+    * @param value the value to insert
+    */
+  def put(key: K, validAt: Long, value: V): Unit = pendingUpdates.synchronized {
+    val competingLatestForKey =
+      pendingUpdates
+        .get(key)
+        .map { pendingUpdate =>
+          val oldLatestValidAt = pendingUpdate.latestValidAt
+          pendingUpdate.latestValidAt =
+            if (pendingUpdate.latestValidAt < validAt) validAt else pendingUpdate.latestValidAt
+          oldLatestValidAt
+        }
+        .getOrElse(Long.MinValue)
+
+    if (competingLatestForKey < validAt) cache.put(key, value) else ()
+  }
+
+  /** Update the cache asynchronously.
+    *
+    * In face of multiple in-flight updates competing for the `key`,
+    * this method registers an async update to the cache
+    * only if the to-be-inserted tuple is the most recent
+    * (i.e. it has `validAt` highest amongst the competing updates).
+    *
+    * @param key the key at which to update the cache
+    * @param validAt ordering discriminator for pending updates for the same key
+    * @param eventualValue the eventual result
+    */
+  final def putAsync(key: K, validAt: Long, eventualValue: Future[V])(implicit
+      loggingContext: LoggingContext
+  ): Future[Unit] = pendingUpdates.synchronized {
+    val pendingUpdatesForKey = pendingUpdates.getOrElseUpdate(key, PendingUpdates.empty)
+    if (pendingUpdatesForKey.latestValidAt < validAt) {
+      pendingUpdatesForKey.latestValidAt = validAt
+      pendingUpdatesForKey.pendingCount += 1
+      registerEventualCacheUpdate(key, eventualValue, validAt)
+    } else Future.unit
+  }
+
+  private def registerEventualCacheUpdate(
+      key: K,
+      eventualUpdate: Future[V],
+      validAt: Long,
+  )(implicit loggingContext: LoggingContext): Future[Unit] =
+    eventualUpdate.transform[Unit](
+      (value: V) =>
+        pendingUpdates.synchronized {
+          if (pendingUpdates(key).latestValidAt == validAt) {
+            cache.put(key, value)
+          }
+          removeFromPending(key)
+        },
+      (err: Throwable) =>
+        pendingUpdates.synchronized {
+          removeFromPending(key)
+          logger.warn(s"Failure in pending cache update for key $key", err)
+          err
+        },
+    )
+
+  private def removeFromPending(key: K): Unit =
+    discard(pendingUpdates.synchronized {
```

Comment:
since it is a private method that is only used from synchronized sections, I would consider removing a redundant intrinsic lock.

Reply:
I've left it there for because the method is not thread-safe and can cause annoying bugs if misused.
But you are right that one synchronize is redundant. I removed it from line 103